frontend/src/components/Chat/ChatWindow.vue

```diff
@@ -0,0 +1,143 @@
+            id="message-scroll"
+          >
+            <template v-slot:default="{ item }">
+              <chat-message
+                :sender="getUsername(item.senderId)"
+                :senderId="item.senderId"
+                :createdAt="item.createdAt"
+                :content="item.content"
+                @chat-message-menu-selection="
+                  $emit('chat-message-menu-selection', $event)
+                "
+              >
+              </chat-message>
+            </template>
+          </v-virtual-scroll>
+        </v-col>
+      </v-row>
+
+      <v-row align="baseline">
+        <v-col class="col-10">
+          <v-text-field
+            v-model="messageContent"
+            label="Message"
+            outlined
+            clearable
+            @keyup.enter="sendMessage"
+          ></v-text-field>
+        </v-col>
+        <v-col class="col-2">
+          <v-btn depressed color="primary" @click="sendMessage"> Send </v-btn>
+        </v-col>
+      </v-row>
+    </v-container>
+  </v-card>
+</template>
+
+<script lang="ts">
+import Vue, { PropType } from 'vue';
+import ChatMessage from '@/components/Chat/ChatMessage.vue';
+import ChatWindowMenu from '@/components/Chat/ChatWindowMenu.vue';
+import { ChannelDto } from '@/common/dto/channel.dto';
+import { MessageDto } from '@/common/dto/message.dto';
+import { UserDto } from '@/common/dto/user.dto';
+
+interface DataReturnType {
+  messageContent: string;
+  itemHeight: number;
+}
+
+export default Vue.extend({
+  props: {
+    channel: {
+      type: Object as PropType<ChannelDto>,
+      required: true,
+    },
+    messages: {
+      type: Map,
+      required: true,
+    },
+    users: {
+      type: Map,
+      required: true,
+    },
+    socket: {
+      type: Object,
+      required: true,
+    },
+  },
+  data(): DataReturnType {
+    return {
+      messageContent: '',
+      itemHeight: 50,
+    };
+  },
+  components: {
+    'chat-message': ChatMessage,
+    'chat-window-menu': ChatWindowMenu,
+  },
+  methods: {
+    printResponse(response: string) {
+      console.log(`Server: ${response}`);
+    },
+    sendMessage() {
+      this.socket.emit('chat-send', {
+        channel: this.channel.id,
+        message: this.messageContent,
+        authorization: this.$store.getters.token,
+      });
+      this.messageContent = '';
+    },
+    scrollDown() {
+      if (this.channel) {
+        let index = (this.messages.get(this.channel.id) as Array<MessageDto>)
+          ?.length;
+        if (index === undefined) index = 0;
+        else index -= 1;
+        const scroller = document.getElementById('message-scroll');
+        if (scroller) {
+          scroller.scrollTop = this.itemHeight * index;
+        }
+      }
+    },
+    getUsername(userId: number): string {
+      const user = this.users.get(userId) as UserDto;
+      if (user) {
+        return user.username;
+      }
+      return 'UNKNOWN USER';
+    },
+    handleLeaveChannelEvent() {
```

Comment:
Don't we need a handleNewMessage as well ? Or Am I missing something ? Right now if a message comes in from another user in the channel currently displayed in the window, the message does not appear right away, and we have to reload the component by switching channels or reloading the page. Tell me if I'm mistaken.

Reply:
This might be the same thing as the unread marker, if you were using the API. It should work if sent from another client on a socket

In production I think messages will only come through sockets, but maybe it is worth implementing the server events just to make testing the frontend easier, because it is probably frustrating having to log-in multiple clients just to see those features

Reply:
Sorry I am just now coming back to this but yes of course, the POST routes don't use sockets... my  bad.
I cannot think of a case where we would want to add a message that did not come from a socket, so I'll stick to opening two clients for now, it's fine.